### PR TITLE
Add Rails/SaveBang cop

### PR DESCRIPTION
**Pull request by QuinnHarris**
Checks for use of save and related Active Record methods where the return type is not handled and the exception raising bang version save! should be used.

This addresses issue https://github.com/bbatsov/rubocop/issues/3128

Bad `model.save`
Good `model.save!` or
```
if model.save
  . . .
else
 . . .
end
```

### CODE REVIEW

File: spec/rubocop/cop/rails/save_bang_spec.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      expect(cop.messages).to be_empty
+    end
+
+    it "when using #{method} with wrong argument" do
+      inspect_source(cop, "object.#{method}('Tom')")
+
+      expect(cop.messages).to be_empty
+    end
+
+    it "when using #{method} with receiver" do
```
Comment (mikegee) [inline spec/rubocop/cop/rails/save_bang_spec.rb]: `object` is the receiver. How about "when assigning the return value of #{method}"?

**Comment by mikegee** [comment]
The build failed because you need to give yourself credit in the changelog. Append this on your line:  `([@QuinnHarris][])`.

**Comment by bbatsov** [comment]
The build is affected by https://github.com/bbatsov/rubocop/issues/3278

**Comment by gerrywastaken** [comment]
@QuinnHarris The 1.9 build issue has been addressed in the latest version of master. However you will still need to add an entry, for your name, to the bottom of the `CHANGELOG`.

If you take a look here you will see that your name isn't being set as a link:
https://github.com/QuinnHarris/rubocop/blob/60136074af6daef872cda6651331d39e7ccbf2cf/CHANGELOG.md

**Comment by QuinnHarris** [comment]
I believe the build and CHANGELOG issues have been addressed in addition to changing a test name.

File: config/disabled.yml [hunk 2]
```diff
@@ -5,6 +5,10 @@
 Rails:
   Enabled: false
 
+Rails/SaveBang:
+  Description: 'Identifies possible cases where Active Record save! or related should be used'
```
Comment (bbatsov) [inline config/disabled.yml]: This should end with a `.`.

File: lib/rubocop/cop/rails/save_bang.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  module Cop
+    module Rails
+      # This cop identifies possible cases where Active Record save! or related
+      # should be used instead of save because the model might have failed to
+      # save and an exception is better than unhandled failure.
+      #
+      # This will ignore calls that are assigned to a variable or used as the
+      # condition in an if/unless statement.  It will also ignore any call with
+      # more than 2 arguments as that is likely not an Active Record call or
+      # if a Model.update(id, attributes) call
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/save_bang.rb]: The final `.` is missing here.

File: lib/rubocop/cop/rails/save_bang.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      #   # good
+      #   unless user.save
+      #      . . .
+      #   end
+      #   user.save!
+      #   user.update!(name: 'Joe')
+      #   user.find_or_create_by!(name: 'Joe')
+      #   user.destroy!
+      class SaveBang < Cop
+        MSG = 'Use %s instead of %s if the return value is not checked.'.freeze
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/save_bang.rb]: Wrap the `%s` in backticks.

File: lib/rubocop/cop/rails/save_bang.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+        SAVE_METHODS = [:save, :create, :update, :destroy,
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/save_bang.rb]: save?

**Comment by bbatsov** [comment]
Apart from my inline remarks - it would be a good idea to add a matching rule to the Rails Style Guide so we can link to it from the cop.

**Comment by QuinnHarris** [comment]
I started a PR on Rails Style Guide https://github.com/bbatsov/rails-style-guide/pull/190
@bbatsov Note sure what the save? remark meant

File: lib/rubocop/cop/rails/save_bang.rb (hunk 5 above)
Comment (mikegee) [inline lib/rubocop/cop/rails/save_bang.rb]: Is there a general term for interacting with the database other than "save"? Maybe "persist".

Comment (bbatsov) [inline lib/rubocop/cop/rails/save_bang.rb]: I don't like the name `SAVE_METHODS`.

**Comment by bbatsov** [comment]
> @bbatsov Note sure what the save? remark meant

I've explained this in an inline comment. The PR has to be squashed and rebased. You can add the link to the new style guide rule as well.

File: spec/rubocop/cop/rails/save_bang_spec.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  shared_examples 'checks_offense' do |method|
+    it "when using #{method} with arguments" do
+      inspect_source(cop, "object.#{method}(name: 'Tom', age: 20)")
+
+      if method == :destroy
+        expect(cop.messages).to be_empty
+      else
+        expect(cop.messages)
+          .to eq([["Use `#{method}!` instead of `#{method}` ",
+                   'if the return value is not checked.'].join])
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/save_bang_spec.rb]: `join` is redundant, btw. You can just have the 2 literals on two adjacent lines and they will automatically be merged into one.

File: lib/rubocop/cop/rails/save_bang.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+          save_loc = node.loc.selector
+          new_method = "#{node.method_name}!"
+
+          ->(corrector) { corrector.replace(save_loc, new_method) }
+        end
+
+        private
+
+        # Ignore simple assignment or if condition
+        def accepted_value_used?(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/save_bang.rb]: Shouldn't `accepted` be `return` here?

File: lib/rubocop/cop/rails/save_bang.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+          return false unless node.parent
+          node.parent.lvasgn_type? ||
+            (node.parent.if_type? && node.sibling_index == 0)
+        end
+
+        # Check argument signature has only one hash
+        def unexpected_argument?(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/save_bang.rb]: I'd name a method like this `expected_signature?` or something like this.

**Comment by bbatsov** [comment]
I made a few extra remarks. Once addressed this will have to be rebased again.

**Comment by arianf** [comment]
There is an issue with this cop: #3439

**Pull request merged**